### DISCUSSION TITLE
Vehicle shape vs ray cast

### DISCUSSION THREAD

**Discussion by LeXXik** [jrouwe/JoltPhysics.js#86] [Q&A] [answered]
Merry Christmas and a Happy New Year!

I am trying to implement the tracked vehicles support and a bit stuck with the spring. Adding a tracked vehicle (e.g. a tank/tractor), I get different results, using the same params. 

Here is a one using rays. Looks good, except the wheels penetrate the step geometry (well, its a ray after all):

https://github.com/jrouwe/JoltPhysics.js/assets/5677782/cff9c74e-dc2a-4f99-b6da-cc818967b167

I would expect cylinder type would help avoiding the clipping, but it behaves very differently (using same spring params for ray/cylinder, wheel radius 0.3, width 0.1, cylinder convex radius fraction 0.3):

https://github.com/jrouwe/JoltPhysics.js/assets/5677782/c14e760d-77b0-4101-bd3b-6a4c42e667a2

**Comment by jrouwe** [comment]
Hello,

I tried to reproduce this in the C++ Samples framework:

https://github.com/jrouwe/JoltPhysics.js/assets/1621693/fff1b4b0-8e78-47c1-b2c1-5251c0409519

https://github.com/jrouwe/JoltPhysics.js/assets/1621693/4e28a22f-2759-4a0c-b10a-d4859c88249e

But it looks good to me, so I have no idea what is different in your setup. Maybe you can share some code?

**Reply by LeXXik** [reply]
Thank you for looking into it! Will try to narrow down the issue on my end.

Is spring stiffness being used when wheel spring is set to `ESpringMode_FrequencyAndDamping`? When I set it to frequency and damping mode, the `spring.HasStiffness()` returns `false`. However, the `stiffness` value is used, e.g. setting it 2 makes the tank bouncy.

**Reply by LeXXik** [reply]
Alright, `HasStiffness` returns true/false if it is valid, not if `stiffness` is used.

**Comment by LeXXik** [comment]
Redacted code on my side (removed unrelated parts and defaults):

```js
const constraintSettings = new Jolt.VehicleConstraintSettings();
const controllerSettings = new Jolt.TrackedVehicleControllerSettings();

const mWheels = constraintSettings.mWheels;
mWheels.clear();
for (let i = 0; i < wheelsCount; i++) {
    const wheel = new Jolt.WheelSettingsTV();
    const spring = wheel.mSuspensionSpring;
    wheel.mPosition = jv.FromBuffer(cb); // jolt vector describing position
    spring.mMode = 0;
    spring.mFrequency = 1;
    spring.mStiffness = 0;
    spring.mDamping = 0;
    mWheels.push_back(wheel);
}

if (!isWheeled) {
    // get tracks and map wheels
    const tracksCount = cb.read(BUFFER_READ_UINT32);
    for (let t = 0; t < tracksCount; t++) {
        const track = controllerSettings.get_mTracks(t);
        const twc = cb.read(BUFFER_READ_UINT32); // track wheels count

        // Make the last wheel in the track to be a driven wheel (connected to engine)
        track.mDrivenWheel = twc - 1;

        for (let i = 0; i < twc; i++) {
            track.mWheels.push_back(cb.read(BUFFER_READ_UINT32)); // wheels indices
        }
    }
}

constraintSettings.mController = controllerSettings;

// constraint
const body = tracker.getBodyByPCID(index);
const constraint = new Jolt.VehicleConstraint(body, constraintSettings);

// wheels contact tester
const tester = new Jolt.VehicleCollisionTesterCastCylinder(1, 0.3);
constraint.SetVehicleCollisionTester(tester);

const listener = new Jolt.VehicleConstraintStepListener(constraint);
physicsSystem.AddConstraint(constraint);
physicsSystem.AddStepListener(listener);
```

**Comment by LeXXik** [comment]
Made a little repro project to see if I can reproduce in the original example. Changing only the tester type doesn't reproduce it - the wheels behave correctly. Must be something else in my code :/ Continue with debugging...


https://github.com/jrouwe/JoltPhysics.js/assets/5677782/385dcf3e-1553-44ab-be4a-9eae1bafb923

**Reply by LeXXik** [reply]
Maybe its just me, not setting the spring defaults correctly. Stiffness shouldn't be zero, regardless of the spring mode. However, I would expect ray and cylinder would behave the same... Also, noticed `stiffness` takes the same value as `frequency`, when not set explicitely. Is this correct?

**Reply by jrouwe** [reply, marked as answer]
Stiffness and frequency are mutually exclusive so they are a C++ union (so assume each others value). They should not be zero. In debug mode this triggers an assert (but you don't see that in JS). If I ignore the assert I can reproduce your issue in the Samples app as well. 

When the frequency is zero, the spring switches to another mode and becomes a constraint that tries to prevent the vehicle from moving along the contact normal (because this code is used for all constraints). The contact normal for the ray vehicle is always pointing up, for a cylinder the normal is not up if it hits the step from the side, this is why it behaves differently.

In any case, setting the frequency to non-zero should fix it.

**Reply by LeXXik** [reply]
I'm also not sure, if I am setting the cylinder tester correctly. For example, the motorcycle example in the repo is using 1 for its radius, while the wheel radius is only 0.31.

https://github.com/jrouwe/JoltPhysics.js/blob/09add344314c9e1721a8b09d2091b8f1fe78e2cf/Examples/vehicle_motorcycle.html#L168

**Reply by jrouwe** [reply]
It's not a radius:

https://jrouwe.github.io/JoltPhysics/class_vehicle_collision_tester_cast_cylinder.html#a416a774b8cf740d496abe37d67d223ff

**Reply by LeXXik** [reply]
Ah, a fraction of a side that becomes a radius, not an actual radius (radius * fraction). I misunderstood it the first time. Thanks!
